modules/payment/commerce_payment.module

```diff
@@ -97,3 +101,66 @@ function template_preprocess_commerce_payment_method(array &$variables) {
     $variables['payment_method'][$key] = $variables['elements'][$key];
   }
 }
+
+/**
+ * Implements hook_form_alter().
+ */
+function commerce_payment_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if ($form_state->get('commerce_checkout_form')) {
+    $form['#validate'][] = 'commerce_payment_process_checkout_payment';
```

Comment:
I almost wonder if checkout should invoke some kind of event for each checkout step validation? "Validate STEP" and allow subscribers to set form state errors.

@bojanz thoughts?